homeassistant/components/asterisk_mbox.py

```diff
@@ -0,0 +1,171 @@
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PORT): int,
+        vol.Required(CONF_PASSWORD): cv.string,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up for the Asterisk Voicemail box."""
+    conf = config.get(DOMAIN)
+
+    host = conf.get(CONF_HOST)
+    port = conf.get(CONF_PORT)
+    password = conf.get(CONF_PASSWORD)
+
+    hass.data[DOMAIN] = AsteriskData(hass, host, port, password)
+
+    discovery.load_platform(hass, "sensor", DOMAIN, {}, config)
+
+    hass.components.frontend.register_built_in_panel(
+        'mailbox', 'Mailbox', 'mdi:account-location')
+    hass.http.register_view(AsteriskMboxMsgView())
+    hass.http.register_view(AsteriskMboxMP3View())
+    hass.http.register_view(AsteriskMboxDeleteView())
+
+    return True
+
+
+class AsteriskData(object):
+    """Store Asterisk mailbox data."""
+
+    def __init__(self, hass, host, port, password):
+        """Init the Asterisk data object."""
+        from asterisk_mbox import Client as asteriskClient
+
+        self.hass = hass
+        self.client = asteriskClient(host, port, password, self.handle_data)
+        self.messages = []
+
+        async_dispatcher_connect(
+            self.hass, SIGNAL_MESSAGE_REQUEST, self._request_messages)
+
+    @callback
+    def handle_data(self, command, msg):
+        """Handle changes to the mailbox."""
+        from asterisk_mbox.commands import CMD_MESSAGE_LIST
+
+        if command == CMD_MESSAGE_LIST:
+            _LOGGER.info("AsteriskVM sent updated message list")
+            self.messages = sorted(msg,
+                                   key=lambda item: item['info']['origtime'],
+                                   reverse=True)
+            async_dispatcher_send(self.hass, SIGNAL_MESSAGE_UPDATE,
+                                  self.messages)
+
+    @callback
+    def _request_messages(self):
+        """Handle changes to the mailbox."""
+        _LOGGER.info("Requesting message list")
+        self.client.messages()
+
+
+class AsteriskMboxMsgView(HomeAssistantView):
+    """View to return the list of messages."""
+
+    url = "/api/asteriskmbox/messages"
+    name = "api:asteriskmbox:messages"
+
+    @asyncio.coroutine
+    def get(self, request):
+        """Retrieve Asterisk messages."""
+        hass = request.app['hass']
+        msgs = hass.data[DOMAIN].messages
+        _LOGGER.info("Sending: %s", msgs)
+        return self.json(msgs)
+
+
+class AsteriskMboxDeleteView(HomeAssistantView):
+    """View to delete selected messages."""
+
+    url = "/api/asteriskmbox/delete"
+    name = "api:asteriskmbox:delete"
+
+    @asyncio.coroutine
+    def post(self, request):
+        """Delete items."""
+        try:
+            data = yield from request.json()
+
+            hass = request.app['hass']
+            client = hass.data[DOMAIN].client
+            for sha in data:
+                _LOGGER.info("Deleting: %s", sha)
+                client.delete(sha)
```

Comment:
Is this doing I/O ? If so, it should not be done inside an async method but be run inside the executor using `yield from hass.async_add_job(client.delete, sha)`